src/cli/subcommands/car_cmd.rs

```diff
@@ -0,0 +1,239 @@
+                let readers: Vec<_> = futures::stream::iter(car_files)
+                    .then(async_fs::File::open)
+                    .map_ok(futures::io::BufReader::new)
+                    .map_err(fvm_ipld_car::Error::from)
+                    .and_then(CarReader::new)
+                    .try_collect()
+                    .await?;
+
+                let all_roots = readers
+                    .iter()
+                    .flat_map(|it| it.header.roots.iter())
+                    .unique()
+                    .cloned()
+                    .collect::<Vec<_>>();
+
+                let car_writer = CarHeader::from(all_roots);
+                let mut output_file =
+                    futures::io::BufWriter::new(async_fs::File::create(output).await?);
+
+                car_writer
+                    .write_stream_async(
+                        &mut output_file,
+                        &mut pin!(dedup_block_stream(merge_car_readers(readers))),
+                    )
+                    .await?;
+            }
+        }
+        Ok(())
+    }
+}
+
+fn read_car_as_stream<R>(reader: CarReader<R>) -> impl Stream<Item = BlockPair>
+where
+    R: AsyncRead + Send + Unpin,
+{
+    futures::stream::unfold(reader, move |mut reader| async {
+        reader
+            .next_block()
+            .await
+            .expect("Failed to call CarReader::next_block")
+            .map(|b| ((b.cid, b.data), reader))
+    })
+}
+
+fn merge_car_readers<R>(readers: Vec<CarReader<R>>) -> impl Stream<Item = BlockPair>
+where
+    R: AsyncRead + Send + Unpin,
+{
+    futures::stream::iter(readers).flat_map(read_car_as_stream)
+}
+
+fn dedup_block_stream(stream: impl Stream<Item = BlockPair>) -> impl Stream<Item = BlockPair> {
+    let mut seen = CidHashSet::default();
+    stream.filter(move |(cid, _data)| futures::future::ready(seen.insert(*cid)))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ahash::HashSet;
+    use cid::multihash;
+    use cid::multihash::MultihashDigest;
+    use cid::Cid;
+    use fvm_ipld_car::Block;
+    use fvm_ipld_encoding::DAG_CBOR;
+    use pretty_assertions::assert_eq;
+    use quickcheck::Arbitrary;
+    use quickcheck_macros::quickcheck;
+
+    #[derive(Debug, Clone)]
+    struct Blocks(Vec<Block>);
+
+    impl From<&Blocks> for HashSet<Cid> {
+        fn from(blocks: &Blocks) -> Self {
+            blocks.0.iter().map(|b| b.cid).collect()
+        }
+    }
+
+    impl Blocks {
+        async fn into_car_bytes(self) -> Vec<u8> {
+            // Dummy root
+            let writer = CarHeader::from(vec![self.0[0].cid]);
+            let mut car = vec![];
+            let mut stream = pin!(futures::stream::iter(self.0).map(|b| (b.cid, b.data)));
+            writer
+                .write_stream_async(&mut car, &mut stream)
+                .await
+                .unwrap();
+            car
+        }
+
+        fn into_stream(self) -> impl Stream<Item = BlockPair> {
```

Comment:
Please change this to a non-owning version.

Reply:
Fixed